lisa/transformers/kernel_source_installer.py

```diff
@@ -0,0 +1,294 @@
+            )
+            result.assert_exit_code()
+
+        make = node.tools[Make]
+        make.make(arguments="olddefconfig", cwd=code_path)
+
+        # set timeout to 2 hours
+        make.make(arguments="", cwd=code_path, timeout=60 * 60 * 2)
+
+    def _install_build_tools(self, node: Node) -> None:
+        os = node.os
+        self._log.info("installing build tools")
+        if isinstance(os, Redhat):
+            os.install_packages(["elfutils-libelf-devel", "openssl-devel", "dwarves"])
+
+            result = node.execute('yum -y groupinstall "Development Tools"', sudo=True)
+            result.assert_exit_code()
+
+            if os.release_version < "8.0.0":
+                # git from default CentOS/RedHat 7.x does not support git tag format
+                # syntax temporarily use a community repo, then remove it
+                node.execute("yum remove -y git", sudo=True)
+                node.execute(
+                    "rpm -U https://centos7.iuscommunity.org/ius-release.rpm", sudo=True
+                )
+                os.install_packages("git2u")
+                node.execute("rpm -e ius-release", sudo=True)
+        elif isinstance(os, Ubuntu):
+            # ccache is used to speed up recompilation
+            # node.execute("command -v ccache", shell=True)
+            # node.execute("export PATH=/usr/lib/ccache:$PATH", shell=True)
+            os.install_packages(
+                [
+                    "git",
+                    "build-essential",
+                    "bison",
+                    "flex",
+                    "libelf-dev",
+                    "libncurses5-dev",
+                    "xz-utils",
+                    "libssl-dev",
+                    "bc",
+                    "ccache",
+                ]
+            )
+        else:
+            raise LisaException(
+                f"os '{os.name}' doesn't support in {self.type_name()}. "
+                f"Implement its build dependencies installation there."
+            )
+
+
+class BaseLocation(subclasses.BaseClassWithRunbookMixin):
+    def __init__(
+        self,
+        runbook: Any,
+        node: Node,
+        parent_log: Logger,
+        *args: Any,
+        **kwargs: Any,
+    ) -> None:
+        super().__init__(runbook, *args, **kwargs)
+        self._node = node
+        self._log = get_logger("kernel_builder", parent=parent_log)
+
+    def get_source_code(self) -> PurePath:
+        raise NotImplementedError()
+
+
+class RepoLocation(BaseLocation):
+    @classmethod
+    def type_name(cls) -> str:
+        return "repo"
+
+    @classmethod
+    def type_schema(cls) -> Type[schema.TypedSchema]:
+        return RepoLocationSchema
+
+    def get_source_code(self) -> PurePath:
+        runbook = cast(RepoLocationSchema, self.runbook)
+        code_path = _get_code_path(runbook.path, self._node, f"{self.type_name()}_code")
+
+        # expand env variables
+        echo = self._node.tools[Echo]
```

Comment:
What's going on here with this echo?

Reply:
as the comments, it expand env variables, like `$HOME`, `~`.